syft/frameworks/torch/mpc/securenn.py

```diff
@@ -305,68 +347,85 @@ def share_convert(a_sh):
-    beta1 = ((a_shares[workers[1].id] + r_shares[workers[1].id]) >= L).long()
-    beta = sy.MultiPointerTensor(children=[beta0, beta1])
+    # a_tilde_sh = a_sh + r_sh
+    # a_shares = a_sh.child
+    # beta0 = (
+    #    ((a_shares[workers[0].id] + r_shares[workers[0].id]) > get_max_val_field(L))
+    #    + ((a_shares[workers[0].id] + r_shares[workers[0].id]) < get_min_val_field(L))
+    # ).type(torch_dtype)
+    # beta1 = (
+    #    ((a_shares[workers[1].id] + r_shares[workers[1].id]) > get_max_val_field(L))
+    #    + ((a_shares[workers[1].id] + r_shares[workers[1].id]) < get_min_val_field(L))
+    # ).type(torch_dtype)
+
+    # beta = sy.MultiPointerTensor(children=[beta0, beta1])
 
     # 4)
-    a_tilde_shares = a_tilde_sh.child
-    delta = (
-        (a_tilde_shares[workers[0].id].copy().get() + a_tilde_shares[workers[1].id].copy().get())
-        >= L
-    ).long()
-    x = a_tilde_sh.get() % L
+    # a_tilde_shares = a_tilde_sh.child
+    # delta = a_tilde_shares[workers[0].id].copy().get() + a_tilde_shares[workers[1].id].copy().get()
+    # Check for both positive and negative overflows
+    # delta = ((delta > get_max_val_field(L)) + (delta < get_min_val_field(L))).type(torch_dtype)
+    # x = a_tilde_sh.get()
 
     # 5)
-    x_bit = decompose(x)
-    x_bit_sh = x_bit.share(*workers, field=p, crypto_provider=crypto_provider, **no_wrap)
-    delta_sh = delta.share(*workers, field=L - 1, crypto_provider=crypto_provider, **no_wrap)
+    # x_bit = decompose(x, L)
+    # x_bit_sh = x_bit.share(
+    #    *workers, field=p, dtype="custom", crypto_provider=crypto_provider, **no_wrap
+    # )
+    # delta_sh = delta.share(
+    #    *workers, field=L - 1, dtype="custom", crypto_provider=crypto_provider, **no_wrap
+    # )
 
     # 6)
-    eta_p = private_compare(x_bit_sh, r - 1, eta_pp)
+    # eta_p = private_compare(x_bit_sh, r - 1, eta_pp, L)
     # 7)
-    eta_p_sh = eta_p.share(*workers, field=L - 1, crypto_provider=crypto_provider, **no_wrap)
+    # eta_p_sh = eta_p.share(
+    #    *workers, field=L - 1, dtype="custom", crypto_provider=crypto_provider, **no_wrap
+    # )
 
     # 9)
-    j = sy.MultiPointerTensor(
-        children=[
-            torch.tensor([0]).send(workers[0], **no_wrap),
-            torch.tensor([1]).send(workers[1], **no_wrap),
-        ]
-    )
-    eta_sh = eta_p_sh + (1 - j) * eta_pp - 2 * eta_pp * eta_p_sh
-
+    # j = sy.MultiPointerTensor(
+    #    children=[
+    #        torch.tensor([0]).send(workers[0], **no_wrap),
+    #        torch.tensor([1]).send(workers[1], **no_wrap),
+    #    ]
+    # )
+    # eta_sh = eta_p_sh + (1 - j) * eta_pp - 2 * eta_pp * eta_p_sh
     # 10)
-    theta_sh = beta - (1 - j) * (alpha + 1) + delta_sh + eta_sh
-
+    # theta_sh = beta - (1 - j) * (alpha + 1) + delta_sh + eta_sh
     # 11)
-    y_sh = -theta_sh + a_sh + u_sh
+    # NOTE:
+    # It seems simple operation with shares in L-1 field is enough to conver a_sh from L to L-1
```

Comment:
No it isn't! This commit introduces subtle bugs in everything that depends on convert share and thus most of SecureNN

Simply try:
```
x = torch.tensor([-2, -1, 0, 1, 2]).share(alice, bob, dtype="custom", field=32, crypto_provider=crypto).child
relu_deriv(x).get() #gives garbage results 
```

Please read carefully the papers; if a mathematician has introduced tons of weird operations, it's probably for some reason.

Reply:
@knexator Could you create an issue for this?

Reply:
@knexator Well we did investigate it quite a bit. But you are right I missed out on it maybe tests I used weren't strict enough. Anyway let's open an issue and work to resolve it quickly.

Reply:
It was left as comments for precisely this reason so that it can be further investigated

Reply:
Also dtype="custom" is not intended for user. I think we need someway to make sure that user can't use custom dtype to share the tensors while at the same time it can be used internally for secureNN ops.

Reply:
#3389

Reply:
@knexator It's alright, no worries :slightly_smiling_face: . Let me know if you need help wrt any changes made here.